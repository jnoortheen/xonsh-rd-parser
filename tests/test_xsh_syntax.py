"""Tests the xonsh parser."""

import pytest

from yaml_snaps import yaml_line_items


@pytest.mark.parametrize(
    "inp, snapped", yaml_line_items("exprs", "stmts"), indirect=["snapped"]
)
def test_line_items(inp, unparse, snapped):
    snapped.matches(unparse(inp))


@pytest.mark.parametrize(
    "inp",
    [
        'x = "WAKKA"; ${x} = 65',
        'x = "."; $(ls @(None or x))',
        'x = "."; !(ls @(None or x))',
        '$[git commit -am "wakka jawaka" ]',
        '$[git commit -am "flock jawaka milwaka" ]',
        '$[git commit -am "wakka jawaka"]',
        '$[git commit -am "flock jawaka"]',
        '![git commit -am "wakka jawaka" ]',
        '![git commit -am "flock jawaka milwaka" ]',
        '![git commit -am "wakka jawaka"]',
        '![git commit -am "flock jawaka"]',
    ],
)
def test_statements(exec_code, inp):
    exec_code(inp, mode="exec")


@pytest.mark.parametrize(
    "inp, result",
    [
        ("$(ls)", ["ls"]),
        ("$(ls )", ["ls"]),
        ("$( ls )", ["ls"]),
        ("$( ls)", ["ls"]),
        ("$(ls .)", ["ls", "."]),
        ('$(ls ".")', ["ls", "."]),
        ("$(ls -l)", ["ls", "-l"]),
        ("$(ls $WAKKA)", ["ls", "wak"]),
        ('$(ls @(None or "."))', ["ls", ["."]]),
        (
            '$(echo hello | @(lambda a, s=None: "hey!") foo bar baz)',
            [["echo", "hello"], (["hey!"], "foo", "bar", "baz")],
        ),
        pytest.param(
            "$(echo @(i**2 for i in range(20) ) )",
            [
                "echo",
                0,
                1,
                4,
                9,
                16,
                25,
                36,
                49,
                64,
                81,
                100,
                121,
                144,
                169,
                196,
                225,
                256,
                289,
                324,
                361,
            ],
            marks=pytest.mark.xfail,
        ),
        pytest.param("$(echo @('a', 7))", ["echo", "a", 7], marks=pytest.mark.xfail),
        pytest.param(
            "$(@$(which echo) ls | @(lambda a, s=None: $(@(s.strip()) @(a[1]))) foo -la baz)",
            "",
            marks=pytest.mark.xfail,
        ),
        ("$(ls $(ls))", ["ls", ["ls"]]),
        ("$(ls $(ls) -l)", ["ls", ["ls"], "-l"]),
        ("$[ls]", ["ls"]),
        ("![ls]", ["ls"]),
        pytest.param(
            "![echo $WAKKA/place]", ["echo", "wak/place"], marks=pytest.mark.xfail
        ),
        ("![echo yo==yo]", ["echo", "yo==yo"]),
        ("!(ls | grep wakka)", [["ls"], ("grep", "wakka")]),
        (
            "!(ls | grep wakka | grep jawaka)",
            [[["ls"], ("grep", "wakka")], ("grep", "jawaka")],
        ),
        ("!(ls > x.py)", ["ls", ">", "x.py"]),
    ],
)
def test_captured_procs(inp, result, exec_code):
    sh = exec_code(inp, xenv={"WAKKA": "wak"})
    assert sh.cmd.result == result

    last_call = sh.cmd.calls[-1]
    match inp[:2]:
        case "$[":
            assert last_call == "run"
        case "$(":
            assert last_call == "out"
        case "![":
            assert last_call == "hide"
        case "!(":
            assert last_call == "obj"


@pytest.mark.parametrize(
    "expr",
    [
        "!(ls)",
        "!(ls )",
        "!( ls)",
        "!( ls )",
        "!(ls .)",
        '!(ls @(None or "."))',
        '!(ls ".")',
        "!(ls $(ls))",
        "!(ls $(ls) -l)",
        "!(ls $WAKKA)",
        "!($LS .)",
    ],
)
def test_bang_procs(expr, exec_code):
    exec_code(expr, xenv={"LS": "ll", "WAKKA": "wak"})


@pytest.mark.parametrize("p", ["", "p"])
@pytest.mark.parametrize("f", ["", "f"])
@pytest.mark.parametrize("glob_type", ["", "r", "g"])
@pytest.mark.xfail
def test_backtick(p, f, glob_type, exec_code):
    exec_code(f"print({p}{f}{glob_type}`.*`)")


@pytest.mark.parametrize(
    "case",
    [
        "![(cat)]",
        "![(cat;)]",
        "![(cd path; ls; cd)]",
        '![(echo "abc"; sleep 1; echo "def")]',
        '![(echo "abc"; sleep 1; echo "def") | grep abc]',
        "![(if True:\n   ls\nelse:\n   echo not true)]",
    ],
)
@pytest.mark.xfail
def test_use_subshell(case, exec_code):
    exec_code(case)


@pytest.mark.parametrize("case", ["", "o", "out", "1"])
@pytest.mark.xfail
def test_redirect_output(case, exec_code):
    assert exec_code(f'$[echo "test" {case}> test.txt]')
    assert exec_code(f'$[< input.txt echo "test" {case}> test.txt]')
    assert exec_code(f'$[echo "test" {case}> test.txt < input.txt]')


@pytest.mark.parametrize("case", ["e", "err", "2"])
@pytest.mark.xfail
def test_redirect_error(case, exec_code):
    assert exec_code(f'$[echo "test" {case}> test.txt]')
    assert exec_code(f'$[< input.txt echo "test" {case}> test.txt]')
    assert exec_code(f'$[echo "test" {case}> test.txt < input.txt]')


@pytest.mark.parametrize("case", ["a", "all", "&"])
@pytest.mark.xfail
def test_redirect_all(case, exec_code):
    assert exec_code(f'$[echo "test" {case}> test.txt]')
    assert exec_code(f'$[< input.txt echo "test" {case}> test.txt]')
    assert exec_code(f'$[echo "test" {case}> test.txt < input.txt]')


@pytest.mark.parametrize(
    "r",
    [
        "e>o",
        "e>out",
        "err>o",
        "2>1",
        "e>1",
        "err>1",
        "2>out",
        "2>o",
        "err>&1",
        "e>&1",
        "2>&1",
    ],
)
@pytest.mark.parametrize("o", ["", "o", "out", "1"])
def test_redirect_error_to_output(r, o, exec_code):
    assert exec_code(f'$[echo "test" {r} {o}> test.txt]')
    assert exec_code(f'$[< input.txt echo "test" {r} {o}> test.txt]')
    assert exec_code(f'$[echo "test" {r} {o}> test.txt < input.txt]')


@pytest.mark.parametrize(
    "r",
    [
        "o>e",
        "o>err",
        "out>e",
        "1>2",
        "o>2",
        "out>2",
        "1>err",
        "1>e",
        "out>&2",
        "o>&2",
        "1>&2",
    ],
)
@pytest.mark.parametrize("e", ["e", "err", "2"])
def test_redirect_output_to_error(r, e, exec_code):
    assert exec_code(f'$[echo "test" {r} {e}> test.txt]')
    assert exec_code(f'$[< input.txt echo "test" {r} {e}> test.txt]')
    assert exec_code(f'$[echo "test" {r} {e}> test.txt < input.txt]')
