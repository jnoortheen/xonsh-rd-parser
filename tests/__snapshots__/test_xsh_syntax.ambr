# serializer version: 1
# name: test_line_items[exprs-bg-0]
  "__xonsh__.cmd('emacs', 'ugggh', bg=True).capture_out()"
# ---
# name: test_line_items[exprs-bg-1]
  "__xonsh__.cmd('emacs', 'ugggh', bg=True).capture_out()"
# ---
# name: test_line_items[exprs-bg-2]
  "__xonsh__.cmd('emacs', 'ugggh', bg=True).capture_out()"
# ---
# name: test_line_items[exprs-bg-3]
  "__xonsh__.cmd('emacs', 'ugggh', bg=True).capture_out()"
# ---
# name: test_line_items[exprs-captured-0]
  "__xonsh__.cmd('cmd', 'sub-cmd', '--opt').capture_out()"
# ---
# name: test_line_items[exprs-captured-1]
  "__xonsh__.cmd('cmd', 'sub-cmd', '--opt').run()"
# ---
# name: test_line_items[exprs-combinators-0]
  "__xonsh__.cmd('ls', '|', 'grep', 'wakka').capture_out()"
# ---
# name: test_line_items[exprs-combinators-1]
  "__xonsh__.cmd('ls', '|', 'grep', 'wakka', '|', 'grep', 'jawaka').capture_out()"
# ---
# name: test_line_items[exprs-env-0]
  "__xonsh__.env['WAKKA']"
# ---
# name: test_line_items[exprs-env-1]
  "__xonsh__.env[None or __xonsh__.env['JAWAKA']]"
# ---
# name: test_line_items[exprs-globs-0]
  "__xonsh__.regex_literal('hello')"
# ---
# name: test_line_items[exprs-globs-10]
  "__xonsh__.cmd('ls', __xonsh__.regex_literal('[Ff]+i*LE'), 'l').capture_out()"
# ---
# name: test_line_items[exprs-globs-1]
  "__xonsh__.cmd('ls', __xonsh__.regex_literal('#[Ff]+i*LE'), 'l').capture_out()"
# ---
# name: test_line_items[exprs-globs-2]
  "__xonsh__.cmd('ls', '[Ff]+i*LE', '-l').capture_out()"
# ---
# name: test_line_items[exprs-globs-3]
  "__xonsh__.cmd('ls', '#[Ff]+i*LE', '-l').capture_out()"
# ---
# name: test_line_items[exprs-globs-4]
  "__xonsh__.cmd('ls', __xonsh__.glob_literal('[Ff]+i*LE'), 'l').capture_out()"
# ---
# name: test_line_items[exprs-globs-5]
  "__xonsh__.cmd('ls', __xonsh__.glob_literal('#[Ff]+i*LE'), 'l').capture_out()"
# ---
# name: test_line_items[exprs-globs-6]
  "__xonsh__.cmd('ls', '@foo`[Ff]+i*LE`', '-l').capture_out()"
# ---
# name: test_line_items[exprs-globs-8]
  "__xonsh__.cmd('ls', '@foo`#[Ff]+i*LE`', '-l').capture_out()"
# ---
# name: test_line_items[exprs-path-literals-0]
  "__xonsh__.path_literal('/foo')"
# ---
# name: test_line_items[exprs-path-literals-10]
  "f'/foo{1 + 1}'"
# ---
# name: test_line_items[exprs-path-literals-11]
  "f'/foo{1 + 1}'"
# ---
# name: test_line_items[exprs-path-literals-12]
  "f'/foo{1 + 1}'"
# ---
# name: test_line_items[exprs-path-literals-1]
  '"""/foo"""'
# ---
# name: test_line_items[exprs-path-literals-2]
  '"""/foo"""'
# ---
# name: test_line_items[exprs-path-literals-3]
  '"""/foo"""'
# ---
# name: test_line_items[exprs-path-literals-4]
  '"""/foo"""'
# ---
# name: test_line_items[exprs-path-literals-5]
  "f'/foo'"
# ---
# name: test_line_items[exprs-path-literals-6]
  "f'/foo'"
# ---
# name: test_line_items[exprs-path-literals-7]
  "f'/foo'"
# ---
# name: test_line_items[exprs-path-literals-8]
  "f'/foo'"
# ---
# name: test_line_items[exprs-path-literals-9]
  "f'/foo{1 + 1}'"
# ---
# name: test_line_items[exprs-redirects-0]
  "__xonsh__.cmd('ls', '>>', 'x.py').capture_object()"
# ---
# name: test_line_items[exprs-redirects-1]
  "__xonsh__.cmd('ls', '|', 'grep', 'wakka', '>', 'x.py').capture_object()"
# ---
# name: test_line_items[exprs-redirects-2]
  "__xonsh__.cmd('ls', '|', 'grep', 'wakka', '>>', 'x.py').capture_object()"
# ---
# name: test_line_items[exprs-redirects-3]
  "__xonsh__.cmd('ls', '>', 'x.py').capture_out()"
# ---
# name: test_line_items[exprs-redirects-4]
  "__xonsh__.cmd('ls', '>>', 'x.py').capture_out()"
# ---
# name: test_line_items[exprs-redirects-5]
  "__xonsh__.cmd('ls', '|', 'grep', 'wakka', '>', 'x.py').capture_out()"
# ---
# name: test_line_items[exprs-redirects-6]
  "__xonsh__.cmd('ls', '|', 'grep', 'wakka', '>>', 'x.py').capture_out()"
# ---
# name: test_line_items[exprs-uncaptured-0]
  "__xonsh__.cmd('cmd', 'sub-cmd', '--opt').run_hidden()"
# ---
# name: test_line_items[exprs-uncaptured-1]
  "__xonsh__.cmd('cmd', 'sub-cmd', '--opt').capture_object()"
# ---
# name: test_line_items[exprs-uncaptured-6]
  "__xonsh__.cmd('echo', '1,2').run_hidden()"
# ---
# name: test_line_items[exprs-uncaptured-7]
  "__xonsh__.cmd('echo', '$foo').capture_object()"
# ---
# name: test_line_items[exprs-uncaptured-8]
  "__xonsh__.cmd('echo', '$foo').capture_object()"
# ---
# name: test_line_items[stmts-assignments-0]
  "__xonsh__.env['WAKKA']"
# ---
# name: test_line_items[stmts-assignments-10]
  '__xonsh__.env[x] = 65'
# ---
# name: test_line_items[stmts-assignments-1]
  "__xonsh__.env['y'] = 'one'"
# ---
# name: test_line_items[stmts-assignments-2]
  "__xonsh__.env['WAKKA'] = 42"
# ---
# name: test_line_items[stmts-assignments-3]
  "y = __xonsh__.env['x']"
# ---
# name: test_line_items[stmts-assignments-4]
  'y = __xonsh__.env[x]'
# ---
# name: test_line_items[stmts-assignments-5]
  "y = __xonsh__.env['x' + 'y']"
# ---
# name: test_line_items[stmts-assignments-6]
  "__xonsh__.env[None or 'WAKKA']"
# ---
# name: test_line_items[stmts-assignments-7]
  "__xonsh__.env[__xonsh__.env['JAWAKA']]"
# ---
# name: test_line_items[stmts-assignments-8]
  "__xonsh__.env['WAKKA'] = 42"
# ---
